### DISCUSSION TITLE
Why does calling an empty function in my Livewire (2.4.4) component mutate the data?

### DISCUSSION THREAD

**Discussion by PhilipVigus** [livewire/livewire#3853] [1. Help]
I have created a simple example to demonstrate the issue.

#### Database
There are two models with a many-to-many relationship. A film can have many tags, and a tag can belong to many films.

#### Livewire
My livewire component consists of the following two files:
```php
class Test extends Component
{
    /** @var Collection */
    public $allTags;

    public function mount()
    {
        $this->allTags = Tag::query()
            ->withCount('films')
            ->orderByDesc('films_count')
            ->orderBy('name')
            ->get()
        ;

        info($this->allTags); // contains the films_count field on each tag
    }

    public function blah()
    {
        info($this->allTags); // films_count is missing
    }

    public function render()
    {
        return view('livewire.test');
    }
}
```
and
```html
<div x-data="{ allTags: {{ $allTags }} }">
    <button wire:click="blah()">Click me</button>

    <template x-for="(tag, index) in allTags" :key="tag.slug">
        <div :index="index" x-text="tag.name + ' x ' + tag.films_count"></a>
    </template>
</div>
```
#### The problem
When I first load the page with the component I get the expected result - A list of tags, each including the number of films that have that tag, eg.
```
Action x 8
Sci Fi x 6
Netflix x 5
```
However, as soon as I click the button (which literally does nothing), the data is mutated and the films_count field is no longer present on the objects in the Livewire component or alpinejs x-data. The page display changes without reloading the page (which I would expect), but looks like this:
```
Action x undefined
Sci Fi x undefined
Netflix x undefined
```
By dumping the contents of allTags on the backend, you can see that film_count is present in allTags at the end of the mount method, but missing in the blah method. This makes no sense to me. The blah method is not updating anything that would cause Livewire to change its state and after mounting, is the only function that is called. But the state is being changed.

Why does calling an empty function in my Livewire component mutate the data, making films_count undefined?

**Comment by ighmouraceneb** [comment]
@PhilipVigus  if you use version v2.6.5 of livewire you just have to replace the mount function by boot

```
   public function boot()
    {
        $this->allTags = Tag::query()
            ->withCount('films')
            ->orderByDesc('films_count')
            ->orderBy('name')
            ->get();
    }
```

if you use the previous version, makes the request in the render() function : 
```
   public function render()
    {
        $this->allTags = Tag::query()
            ->withCount('films')
            ->orderByDesc('films_count')
            ->orderBy('name')
            ->get();

            return view(...);
    }
```
or pass $allTags as a parameter to the blah($allTags) function, and reset the $this->$allTags variable

```
<div x-data="{ allTags: {{ $allTags }} }">
    <button wire:click="blah($allTags)">Click me</button>

    <template x-for="(tag, index) in allTags" :key="tag.slug">
        <div :index="index" x-text="tag.name + ' x ' + tag.films_count"></a>
    </template>
</div>
```

```
class Test extends Component
{
    /** @var Collection */
    public $allTags;

    public function mount()
    {
        $this->allTags = Tag::query()
            ->withCount('films')
            ->orderByDesc('films_count')
            ->orderBy('name')
            ->get()
        ;

        info($this->allTags); // contains the films_count field on each tag
    }

    public function blah($allTags)
    {
       $this->allTags = $allTags
        info($this->allTags); // films_count is missing
    }

    public function render()
    {
        return view('livewire.test');
    }
}
```
 [for more information on the boot and mount functions](https://laravel-news.com/livewire-v2-6-0)

**Reply by PhilipVigus** [reply]
@ighmouraceneb Thanks for the detailed answer. I guess I have a follow up question, which is _why_ Livewire does this? It would be really helpful to understand why films_count disappears in this way, but the rest of the $allTags collection stays the same. Intuitively I would have expected not to have to use any of the techniques you suggest?

**Reply by jdesousa1** [reply]
> @ighmouraceneb Thanks for the detailed answer. I guess I have a follow up question, which is why Livewire does this? It would be really helpful to understand why films_count disappears in this way, but the rest of the $allTags collection stays the same. Intuitively I would have expected not to have to use any of the techniques you suggest?

Hi, I have this problem right now. I don't have a solution but I know something.

If you see the query params on each request, livewire pass data like:

```
{
    //...
    dataMeta: {
        models: {
            model: {
                class: "App\\Models\\Model",
                connection: "mysql",
                id: 1,
                relations: ["relation"]
            }
        }
    }
}
```

So... I understand that livewire make a new request on database with this params. It works with "with()" method but doesn't work "withCount()". First time you data is good, but when you have interaction with livewire, it make a new request without withCount ('films').

You can see that using dev tools like clockwork or debugbar.

Meanwhile, I'll keep looking for a good solution because I don't want to make a new request on database for a single action that it doesn't have relation.
